### PR TITLE
P2022_1267 Create Recycler View

### PR DESCRIPTION
**Pull request by michalfiuk1**
# P2022_1267

## PROBLEM
As a user, I would like to be able to browse available products in the app. See primary information like product name image and its price.

AC:

- Product list is displayed and scrollable
- Each item contains an image (update mock data source if needed), the title of the product, and the price
- Product without image displays placeholder
- Upon clicking the item toast message appears with the message "Not Implemented [P2022-630](https://tracker.intive.com/jira/browse/P2022-630)"

## SOLUTION
Add working URL for the image in products.json from the Internet and in AndroidManifest.xml allow connecting Internet for loading images. Also, add a placeholder for unloaded images.
New argument (imageURL) in ProductListMapper.kt and Product.kt.
Change return in CategoriesViewModel.kt to LiveData(fix).
Make Home fragment as a recycler view for showing products (Create HomeViewModel,  HomeListAdapter, and home_single_item.xml).
Start using glide to handle images. Add dependence in build.gradle for glide and string for description images in strings.xml.

## SCREENSHOTS
After:
image
Before:
image

## SIDE NOTES
[P2022-1268](https://tracker.intive.com/jira/browse/P2022-1268) 
this is included in the subtask. Removed Hello Hello from activity_main.xml.

### CODE REVIEW

File: app/src/main/res/layout/home_single_item.xml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
```
Comment (rafal-depta-intive) [inline app/src/main/res/layout/home_single_item.xml:6]: here you can use `wrap-content` for layout_height, so that the elements don't take up the whole screen

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -19,5 +19,6 @@
     <string name="settings">Settings</string>
     <string name="about">About</string>
     <string name="hello_hello">Hello Hello</string>
+    <string name="product_image_description">product_image</string>
```
Comment (rafal-depta-intive) [inline app/src/main/res/values/strings.xml:22]: For string value we are typing "normal" tekst starting with a capital letter, so here it will be `Product image`

File: app/src/main/java/com/intive/patronage22/lublin/MainActivity.kt [hunk 3]
```diff
@@ -20,6 +20,7 @@ import com.intive.patronage22.lublin.screens.search.SearchFragment
 
 class MainActivity : AppCompatActivity() {
 
+
```
Comment (rafal-depta-intive) [inline app/src/main/java/com/intive/patronage22/lublin/MainActivity.kt:23]: Is this change necessary?

**Comment by rafal-depta-intive** [comment]
As I can see on screens you attached that there is text `Hello Hello` on Home screen, I think we can remove that

File: app/src/main/java/com/intive/patronage22/lublin/screens/home/HomeFragment.kt [hunk 4]
```diff
@@ -1,8 +1,43 @@
...
+        val view = inflater.inflate(R.layout.fragment_home, container, false)
+        val productListView = view.findViewById<RecyclerView>(R.id.listProducts)
+        viewModel.products.observe(viewLifecycleOwner) { products ->
+            productListView.adapter = HomeListAdapter(this, products)
+        }
+        return view
+    }
+
+
+
```
Comment (dziubinski-intive) [inline app/src/main/java/com/intive/patronage22/lublin/screens/home/HomeFragment.kt:42]: empty lines

File: app/src/main/java/com/intive/patronage22/lublin/screens/home/HomeFragment.kt [hunk 5]
```diff
@@ -1,8 +1,43 @@
...
+            )
+    }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        val view = inflater.inflate(R.layout.fragment_home, container, false)
+        val productListView = view.findViewById<RecyclerView>(R.id.listProducts)
```
Comment (dziubinski-intive) [inline app/src/main/java/com/intive/patronage22/lublin/screens/home/HomeFragment.kt:34]: https://developer.android.com/topic/libraries/view-binding

File: app/src/main/java/com/intive/patronage22/lublin/screens/home/HomeListAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                .load(product.imageUrl)
+                .placeholder(R.drawable.image_placeholder)
+                .fitCenter()
+                .into(productImageView)
+        }
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(parent.context)
+            .inflate(R.layout.home_single_item, parent, false)
```
Comment (dziubinski-intive) [inline app/src/main/java/com/intive/patronage22/lublin/screens/home/HomeListAdapter.kt:35]: Use https://developer.android.com/topic/libraries/view-binding

and pass Binding to the holder

**Comment by michalfiuk1** [comment]
Can we add a subtask to the appearance of the application to this sprint so as not to throw everything in one task?

**Comment by dziubinski-intive** [comment]
Yeah, definitely agree, you can address it separately. As long as ticket didn't specify much regarding UI it's totally fine to have it handled in even in next sprint.

File: app/src/main/java/com/intive/patronage22/lublin/screens/categories/CategoriesListAdapter.kt [hunk 7]
```diff
@@ -9,27 +9,37 @@ import android.widget.TextView
...
+    private val fragment: CategoriesFragment,
+    private val categories: List<String>
+) :
     RecyclerView.Adapter<CategoriesListAdapter.ViewHolder>() {
 
-    class ViewHolder(private val context: Context, itemView: View) :
-        RecyclerView.ViewHolder(itemView) {
-        private val categoryTextView: TextView = itemView.findViewById(R.id.categoriesListText)
+    class ViewHolder(
+        private val fragment: CategoriesFragment,
```
Comment (dziubinski-intive) [inline app/src/main/java/com/intive/patronage22/lublin/screens/categories/CategoriesListAdapter.kt:21]: it's not needed to pass whole fragment here. You can grab context `binding.root`

**Comment by dziubinski-intive** [comment]
Just added smal remark, you can address it here on in following-up PR.

**Review by dziubinski-intive — APPROVED** [review]
